man/man1/pcp-kube-pods.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2018 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PCP_KUBE_PODS 1 "PCP" "Performance Co-Pilot"
.ds xM PCP_KUBE_PODS
.SH NAME
\f3pcp-kube-pods\f1 \- list Kubernetes pods to scan for running PCP services
.SH SYNOPSIS
\f3$PCP_BINADM_DIR/discover/pcp-kube-pods\f1
.SH DESCRIPTION
.B pcp-kube-pods
uses
.BR kubectl (1)
to provide a list of IP addresses for PODs running in a local
Kubernetes cluster, that may be running PCP services like
.BR pmcd (1)
and
.BR pmproxy (1).
.PP
It is used by the
.BR pmfind (1)
command and the
.BR pmDiscoverServices (3)
API as a ``shell'' command.
.PP
The script invokes the
.B "kubectl get pod"
command line (see
.BR kubectl-get (1))
to discover IP addresses for pods.
The
.B .status.podIP
output field is extracted from the Kubernetes pod object(s).
Additional options can be specified via configuration file,
such as
.I "-l service=database"
to restrict the results using Kubernetes pod labels.
The default configuration file is
.BR $PCP_SYSCONF_DIR/discover/pcp-kube-pods.conf .
.PP
If no local
.B kubectl
command is found, nothing is reported and an exit code
indicating success is returned.
.SH DIAGNOSTICS
All diagnostics produced are provided by the
.B kubectl
command.
.P
The
.BR kubectl
exit status is propagated unchanged.
.SH FILES
.TP 5
.I $PCP_BINADM_DIR/discover
default path to address discovery commands like
.B pcp-kube-pods
.TP
.I $PCP_SYSCONF_DIR/discover/pcp-kube-pods.conf
configuration file containing command line options passed to
.B kubectl-get
.TP
.I $PCP_SYSCONF_DIR/discover/pcp-kube-pods.disabled
existence of this file causes the script to report nothing
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR kubectl-get (1),
.BR kubectl (1),
.BR pmcd (1),
.BR pmfind (1),
.BR pmDiscoverServices (3)
and
.BR pcp.env (5).

.\" control lines for scripts/man-spell
.\" +ok+ Kubernetes PCP_KUBE_PODS PODs kubectl podIP


CHANGELOG
pcp-7.1.0 (4 February 2026)
    - Note: Ubuntu Resolute feature freeze is in February
    - Work-in-progress.

pcp-7.0.5 (21 November 2025)
    - build: disable usdt.h use in configure script on armv8l (Debian)

pcp-7.0.4 (20 November 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdasmart: Support for collecting NVMe error log (0x01) entries
	pmdadarwin: add kernel.all.distro metric code
    - Build, infrastructure, containers and packaging updates:
	build: check linking with usdt.h in configure script, not just compiling
	build: rework debian post* and pre* scripts
	build: rework migration from misnamed *t64* packages on Ubuntu
	vendor: update htop to latest upstream

pcp-7.0.3 (12 November 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdafarm: extend support for SCSI (SAS) FARM log
	pmdafarm: correct help text lengths, missing indom help text
	pmdafarm: update pmUnits for metrics
	pmdainfiniband: add collection of additional metrics
	pmdainfiniband: updated metrics help text, semantics and units
	pmdadm: add support for collecting dmcrypt metrics
	pmdadm: add dmcrypt indom help text
	pmdahdb: fix selinux issue
	pmdahdb: graceful fail when connection fails
	pmdads389: add timeouts on connect/bind and fetch/search
	pmdabpf: switch to using local-built library and headers
	pmdabpf: updates to opensnoop and mountsnoop to match kernel
	pmdanetcheck: updates for Python 3.14
	pmdaopenmetrics: fix performance bug in refresh
	pmdaopentelemetry: pmns refresh fix, same as for openmetrics
	pmdaproc: fix process hog metrics
    - Client tools and utilities:
	pcp-dstat: add a vLLM plugin showing cache utilisation
	pcp-uptime: assorted logic corrections
	pcp2ganglia: make the build conditional on the Perl RRDs package
	guidellm2pcp: updates for upcoming version (0.4) format changes
    - Server-side utilities and log management scripts:
	pmlogctl, pmiectl: refactor common code into a common script
	pmlogctl: fix intermittent qa/1483 failure relating to a diagnostic
	pmlogconf: clarify scope rules for name in define control lines
	pmlogconf: add support for (new) define control line
	pmlogconf: add configuration for CIFS shares logging
	pmlogconf: add configuration for FARM metrics logging
	pmlogconf: add configuration for SMART metrics logging
	pmlogconf: add collection of gfs2.glockfd metrics in base config
	pmlogger: do not register SIG_DFL as signal handler (valgrind)
	pmcd: generate bpftrace script, improve tracing output
	pmcd: transition to simpler libbpf project usdt.h and macros
	pmproxy: fix null pointer dereference
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: change __pmGetAddrInfo() API for improved error handling
	libpcp: add pmUnregisterDerived() API
	libpcp: remove local strchrnul definition entirely
	libpcp: macOS getaddrinfo requires extra non-failure check
	libpcp_web: report NULL pointer issues as errors instead of debug
	libpcp_web: add checks for NULL pointer dereference during discovery
	libpcp_web: add checks for NULL pointer when context is released
	libpcp_web: tweak recent diagnostic update for uniqueness
	libpcp_web: remove compiler warning
	python: extend get_indom() to handle int as argument
	python: replace strftime('%s') uses with timestamp()
	python: pylint cleanup in pcp-uptime
	pmieconf: switch to using termios API for macOS builds
	pmlogrewrite: rework use of assert()
    - Build, infrastructure, containers and packaging updates:
	build: drop ppc64el from deb platforms with bpfcc
	build: bump minimum clang version for BPF PMDA
	build: deal with debian12 inability to compile bash bpf module
	build: drop final reference to long-retired termio.h, fixing f43
	build: try reorder static library linkage for Debian platforms
	build: debian packaging list updates for ibverbs
	build: update debian infiniband package requirements
	build: fix version number in redhat.spec file
	build: resolve CI failure on macOS with latest toolchain
	ci: fedora43 is released now, add to automation
	vendor: update htop to latest upstream
	vendor: update libbpf, bpftool, usdt and bcc libraries
	package lists: updates for Debian 13, CentOS Stream 10, openSUSE Leap 16.0
	package lists: updates for FreeBSD, OpenBSD 7.8
    - Documentation:
	docs: switch from Mac OS X to macOS

pcp-7.0.2 (15 October 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: update metric removal mechanism
	pmdaopentelemetry: QA failure fixes and improvements
	pmdadocker: auto detect PCP_DOCKER_DIR using the Docker Engine API
	pmdahacluster: support "cibadmin --local" deprecation for pacemaker metrics
	pmdasmart: Better handling of non-supported NVME features
	pmdanvidia: add pmlogconf setups for logging GPU metrics
	pmdacifs: add support for collecting DebugData metrics
	pmdahdb: remove unused SPS version macros, various improvements
	pmdalinux: add new numastat per-node metrics (mem.numa.util.*)
	pmdarocestat: fix invalid path on RHEL8, support MLNX OFED 4.1
	pmdaproc: fix macOS ticks calculation for process time
	pmdaproc: updates to the macOS scheduler state metric
    - Client tools and utilities:
	pcp-meminfo: fix handling of hinv.hugepagesize which uses bytes
	pcp-atop: fix column width for CID when showing GPU stats (e)
	pcp-dstat: reduce width of GPU columns, more like the CPU columns
	pcp-dstat: fix incremental sample updating, add default time column
	pcp-xsos: source code merged from separate xsos-pcp repository
	pcp-htop: sync with latest htop.dev code fixes
	vllmbench2pcp: convert vLLM bench results to PCP archives
	guidellm2pcp: add run_id as a metric also, for ease of analysis
	guidellm2pcp: ensure instance identifier stability across runs
    - Server-side utilities and log management scripts:
	pmlogctl: fix variable renaming fallout
	pmlogconf: insert additional metrics needed for pcp-numastat
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix formatting arguments to xz_debug()

src/discover/pcp-kube-pods.sh
#!/bin/sh
#
# Copyright (c) 2018,2020 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
. $PCP_DIR/etc/pcp.env

PCP_KUBECTL_PROG=${PCP_KUBECTL_PROG-'kubectl'}
which $PCP_KUBECTL_PROG >/dev/null 2>/dev/null || exit 0
[ -e "$PCP_SYSCONF_DIR/discover/pcp-kube-pods.disabled" ] && exit 0

args="-o jsonpath={.items[*].status.podIP}"
file="$PCP_SYSCONF_DIR/discover/pcp-kube-pods.conf" 
[ -f "$file" ] && args=`cat "$file"`

exec $PCP_KUBECTL_PROG get pods $args
